Assets/Player/CameraControls.gd

```diff
@@ -74,18 +71,45 @@ func _move_drag() -> void:
 			_origin.translate(move_dir)
 			_drag_pos = new_drag_pos
 
-func _rotate(rotation: float) -> void:
+func rotate(rotation: float) -> void:
 	_rotation_y.rotate_y(rotation)
 	_basis = _get_basis()
 
-func _zoom(zoom_value: float) -> void:
-	var m_pos := _viewport.get_mouse_position()
-	var start_result := _raycast_from_mouse(m_pos, 1)
+func _is_zoom_within_limit(zoom_value: float) -> bool:
+	return (zoom_value < 0 and _camera.size > ZOOM_IN_LIMIT) or (zoom_value > 0 and _camera.size < ZOOM_OUT_LIMIT)
+
+func _zoom_limit_check():
+	Global.emit_signal("camera_max_zoom", _camera.size <= ZOOM_IN_LIMIT)
+	Global.emit_signal("camera_min_zoom", _camera.size >= ZOOM_OUT_LIMIT)
+
+func _zoom_mouse(zoom_value: float) -> void:
+	if _is_zoom_within_limit(zoom_value):
+		var m_pos := _viewport.get_mouse_position()
+		var start_result := _raycast_from_mouse(m_pos, 1)
+		_zoom_camera(zoom_value)
+		var end_result := _raycast_from_mouse(m_pos, 1)
+		if not (start_result.is_empty() and end_result.is_empty()):
+			var move_dir: Vector3i = start_result.position - end_result.position
+			_origin.translate(move_dir)
+
+		_zoom_limit_check()
```

Comment:
Please read https://docs.godotengine.org/en/stable/tutorials/physics/ray-casting.html#d-ray-casting-from-screen

Also you can use math instead physics engine (Pythagore). You know your plane is always on Y = 0.
Not same, but can help you https://github.com/theludovyc/Godot_DestropolisLike/blob/1081bde949fd27e9d43eb0d3e4cfa5e36faa1b8a/Script/Camera3D.gd#L15